bot/exts/halloween/spookynamerate.py

```diff
@@ -0,0 +1,409 @@
+                msg = await channel.fetch_message(message_id)
+                score = 0
+                for reaction in msg.reactions:
+                    reaction_value = EMOJIS_VAL.get(reaction.emoji, 0)  # get the value of the emoji else 0
+                    score += reaction_value * (reaction.count - 1)  # multiply by the num of reactions
+                    # subtract one, since one reaction was done by the bot
+
+                logger.debug(f"{self.bot.get_user(data['author'])} got a score of {score}")
+                data["score"] = score
+                await self.messages.set(message_id, json.dumps(data))
+
+            # Sort the winner messages
+            winner_messages = sorted(
+                ((msg_id, json.loads(usr_data)) for msg_id, usr_data in await self.messages.items()),
+                key=lambda x: x[1]["score"],
+                reverse=True,
+            )
+
+            winners = []
+            for i, winner in enumerate(winner_messages):
+                winners.append(winner)
+                if len(winner_messages) > i + 1:
+                    if winner_messages[i + 1][1]["score"] != winner[1]["score"]:
+                        break
+                elif len(winner_messages) == (i + 1) + 1:  # The next element is the last element
+                    if winner_messages[i + 1][1]["score"] != winner[1]["score"]:
+                        break
+
+            # one iteration is complete
+            await channel.send("Today's Spooky Name Rate Game ends now, and the winner(s) is(are)...")
+
+            async with channel.typing():
+                await asyncio.sleep(1)  # give the drum roll feel
+
+                if not winners:  # There are no winners (no participants)
+                    await channel.send("Hmm... Looks like no one participated! :cry:")
+                    return
+
+                if len(winners) > 1:
+                    await channel.send(" and ".join(PING.format(id=win[1]["author"]) for win in winners) + " !")
+                    score = winners[0][1]["score"]
+                    # display winners, their names and scores
+                    await channel.send(f"Congratulations to all! You have a score of {score}!")
+                    names = ", ".join([f'{win[1]["name"]} ({PING.format(id=win[1]["author"])})' for win in winners])
+                    await channel.send(f"Your names were: {names}!")
+
+                else:  # there is only one winner
+                    # There cant be 0 winners since we are checking that above
+
+                    winner = winners[0][1]
+                    await channel.send(f"{PING.format(id=winner['author'])}!")
+                    await channel.send(
+                        f"Congratulations {PING.format(id=winner['author'])}! You have a score of "
+                        f"{winner['score']}!"
+                    )
+                    await channel.send(f"Your name was: **{winner['name']}**.")  # display winner, name and score
```

Comment:
This can be combined into 1. You can re-use the join logic for singular winners as well, since it wont add the join string to a single winner. You just need a condition for the difference in text, like: (Also don't need the temporary list)
```py
names = ", ".join(f'{win[1]["name"]} ({PING.format(id=win[1]["author"])})' for win in winners)
congratulations = 'to all' if len(winners) > 1 else PING.format(id=winners[0][1]['author'])
score = winners[0][1]["score"]
await channel.send(
    f"Congratulations { congratulations }!\n"
    f"You have a score of { score }\n"
    f"Your name{ 's were' if len(winners) > 1 else 'was' }: **{ names }**!"
)
```
Give or take a bit, just a rough sketch of what it could be.
Point is, there's a lot duplicated here, and the joins don't actually need to seperated out of 1 or multiple winners.
Additionally I don't like sending 3 messages instantly just to list the names/winners. (which will also ping people twice..)